### PR TITLE
feat: add context to the server and requests

### LINKED ISSUE #7: [server] Add generic context
When creating a server, we need a context to be shared and it should be generic.

### PR DESCRIPTION
**Pull request by lauti7**
This PR: 
- add context to the server
- change the example to have context

Closes #7

### CODE REVIEW

File: examples/integration/service/book_service.rs [hunk 1]
```diff
@@ -1,15 +1,32 @@
...
+
+        let book = ctx
+            .hardcoded_database
+            .iter()
+            .find(|book_record| book_record.isbn == request.isbn);
+
+        let mut final_book = Book::default();
+
+        if book.is_some() {
+            let b = book.unwrap();
```
Comment (moliva) [inline examples/integration/service/book_service.rs:23]:
```suggestion
        if let Some(b) = book {
```

File: examples/integration/service/book_service.rs [hunk 2]
```diff
@@ -1,15 +1,32 @@
...
+            final_book.set_author(b.author.clone());
+            final_book.set_isbn(b.isbn.clone());
+            final_book.set_title(b.title.clone());
+            return final_book;
```
Comment (moliva) [inline examples/integration/service/book_service.rs:27]: This line seems to be redundant and adds complexity to the flow of the logic.

File: examples/integration/service/book_service.rs [hunk 3]
```diff
@@ -1,15 +1,32 @@
...
+        }
+
+        final_book
```
Comment (moliva) [inline examples/integration/service/book_service.rs:30]: Another version of the above block could be:
```rust
book.map(Book::clone).unwrap_or_default()
```

File: examples/integration/codegen/mod.rs [hunk 4]
```diff
@@ -1,25 +1,28 @@
...
+use crate::{
+    service::api::{Book, GetBookRequest},
+    MyExampleContext,
+};
 
 pub const SERVICE: &str = "BookService";
 
 pub trait BookServiceInterface {
-    fn get_book(&self, request: GetBookRequest) -> Book;
+    fn get_book(&self, request: GetBookRequest, ctx: &MyExampleContext) -> Book;
```
Comment (guidota) [inline examples/integration/codegen/mod.rs:12]:
```suggestion
    fn get_book(&self, request: GetBookRequest, context: &MyExampleContext) -> Book;
```

File: examples/integration/codegen/mod.rs [hunk 5]
```diff
@@ -1,25 +1,28 @@
...
-        port: &mut RpcServerPort,
+        port: &mut RpcServerPort<MyExampleContext>,
         service: S,
     ) {
         println!("> BookServiceCodeGen > register_service");
         let mut service_def = ServiceModuleDefinition::new();
-        service_def.add_definition("GetBook".to_string(), move |request| {
-            let res = service.get_book(GetBookRequest::parse_from_bytes(request).unwrap());
+        service_def.add_definition("GetBook".to_string(), move |request, ctx| {
+            let res = service.get_book(GetBookRequest::parse_from_bytes(request).unwrap(), ctx);
```
Comment (guidota) [inline examples/integration/codegen/mod.rs:25]:
```suggestion
        service_def.add_definition("GetBook".to_string(), move |request, context| {
            let res = service.get_book(GetBookRequest::parse_from_bytes(request).unwrap(), context);
```

File: src/server.rs [hunk 6]
```diff
@@ -17,7 +17,7 @@ use crate::{
     },
 };
 
-type PortHandlerFn = dyn Fn(&mut RpcServerPort) + Send + Sync + 'static;
+type PortHandlerFn<CTX> = dyn Fn(&mut RpcServerPort<CTX>) + Send + Sync + 'static;
```
Comment (guidota) [inline src/server.rs:20]:
```suggestion
type PortHandlerFn<Context> = dyn Fn(&mut RpcServerPort<Context>) + Send + Sync + 'static;
```

File: examples/integration/service/book_service.rs (hunk 3 above)
Comment (lauti7) [reply examples/integration/service/book_service.rs:30]: great! After changing some types for `book` it worked

**Pull request merged**